File: src-tauri/tests/fixtures/echo_mcp_server.py
#!/usr/bin/env python3
"""Minimal MCP stdio server for integration tests.

Exposes a single `echo` tool that returns its input.
"""
import json
import sys


def send_response(req_id, result):
    response = {"jsonrpc": "2.0", "id": req_id, "result": result}
    sys.stdout.write(json.dumps(response) + "\n")
    sys.stdout.flush()


def main():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            msg = json.loads(line)
        except json.JSONDecodeError:
            continue

        method = msg.get("method")
        req_id = msg.get("id")

        if method == "initialize":
            send_response(
                req_id,
                {
                    "protocolVersion": "2024-11-05",
                    "capabilities": {},
                    "serverInfo": {"name": "echo", "version": "1.0.0"},
                },
            )
        elif method == "tools/list":
            send_response(
                req_id,
                {
                    "tools": [
                        {
                            "name": "echo",
                            "description": "Echoes the provided message",
                            "inputSchema": {
                                "type": "object",
                                "properties": {
                                    "message": {"type": "string"}
                                },
                                "required": ["message"],
                            },
                        }
                    ]
                },
            )
        elif method == "tools/call":
            args = msg.get("params", {}).get("arguments", {})
            message = args.get("message", "")
            send_response(
                req_id,
                {
                    "content": [
                        {"type": "text", "text": f"echo: {message}"}
                    ]
                },
            )


if __name__ == "__main__":
    main()

File: src-tauri/tests/fixtures/echo_mcp_server.rs
use std::io::{BufRead, Write};

fn send_response(id: u64, result: serde_json::Value) {
    let response = serde_json::json!({
        "jsonrpc": "2.0",
        "id": id,
        "result": result,
    });
    println!("{}", serde_json::to_string(&response).unwrap());
    std::io::stdout().flush().unwrap();
}

fn main() {
    let stdin = std::io::stdin();
    for line in stdin.lock().lines() {
        let line = match line {
            Ok(line) => line,
            Err(_) => break,
        };
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let msg: serde_json::Value = match serde_json::from_str(line) {
            Ok(msg) => msg,
            Err(_) => continue,
        };

        let method = msg.get("method").and_then(|m| m.as_str()).unwrap_or("");
        let id = msg.get("id").and_then(|i| i.as_u64()).unwrap_or(0);

        match method {
            "initialize" => {
                send_response(
                    id,
                    serde_json::json!({
                        "protocolVersion": "2024-11-05",
                        "capabilities": {},
                        "serverInfo": {"name": "echo", "version": "1.0.0"},
                    }),
                );
            }
            "tools/list" => {
                send_response(
                    id,
                    serde_json::json!({
                        "tools": [
                            {
                                "name": "echo",
                                "description": "Echoes the provided message",
                                "inputSchema": {
                                    "type": "object",
                                    "properties": {
                                        "message": {"type": "string"}
                                    },
                                    "required": ["message"],
                                },
                            }
                        ]
                    }),
                );
            }
            "tools/call" => {
                let args = msg
                    .get("params")
                    .and_then(|p| p.get("arguments"))
                    .cloned()
                    .unwrap_or_else(|| serde_json::json!({}));
                let message = args
                    .get("message")
                    .and_then(|m| m.as_str())
                    .unwrap_or("")
                    .to_string();
                send_response(
                    id,
                    serde_json::json!({
                        "content": [
                            {"type": "text", "text": format!("echo: {message}")}
                        ]
                    }),
                );
            }
            _ => {}
        }
    }
}
